File: flowersec-swift/Sources/Flowersec/RetryDisposition.swift
/// The structured retry decision used by ``ConnectionController``.
///
/// Retry deadlines use exact Unix milliseconds on every platform so the
/// contract does not depend on `Date` precision or rounding.
public enum RetryDisposition: Equatable, Sendable {
  case terminal
  case retryable
  case retryAfter(UInt64)
}

extension SessionError {
  public var retryDisposition: RetryDisposition {
    switch self {
    case .canceled, .streamRejected, .operationFailed:
      return .terminal
    case .timeout, .closed, .goingAway, .resourceExhausted, .streamReset, .rekeyFailed,
      .livenessFailed:
      return .retryable
    }
  }
}

File: flowersec-ts/src/v3/retryDisposition.ts
import type { SessionErrorCode } from "./contract.js";
import type { PublicConnectErrorCodeV3 } from "./security.js";

export type RetryDisposition =
  | Readonly<{ kind: "terminal" }>
  | Readonly<{ kind: "retryable" }>
  | Readonly<{ kind: "retry_after"; notBeforeUnixMilliseconds: number }>;

export function retryDispositionForConnectError(
  error: Readonly<{ code: PublicConnectErrorCodeV3 }>,
): RetryDisposition {
  switch (error.code) {
    case "expired_artifact":
    case "connection_failed":
      return Object.freeze({ kind: "retryable" });
    default:
      return Object.freeze({ kind: "terminal" });
  }
}
export function retryDispositionForSessionError(error: Readonly<{ code: SessionErrorCode }>): RetryDisposition {
  switch (error.code) {
    case "closed":
    case "going_away":
    case "timeout":
    case "resource_exhausted":
    case "stream_reset":
    case "rekey_failed":
    case "liveness_failed":
      return Object.freeze({ kind: "retryable" });
    default:
      return Object.freeze({ kind: "terminal" });
  }
}
